conformance/packages/dns-test/src/docker/hickory.Dockerfile

```diff
@@ -1,23 +1,39 @@
-FROM rust:1-slim-bookworm
-
+FROM rust:1-slim-bookworm AS chef
+ENV CARGO_INCREMENTAL=0
+ENV CARGO_PROFILE_DEV_DEBUG=0
+ENV CARGO_PROFILE_DEV_STRIP=true
+RUN cargo install cargo-chef --version 0.1.71 --profile dev
 ARG DNSSEC_FEATURE=dnssec-ring
 
-# ldns-utils = ldns-{key2ds,keygen,signzone}
-RUN apt-get update && \
-    apt-get install -y \
-        ldnsutils \
-        bind9-utils \
-        tshark \
-        libssl-dev \
-        pkg-config
-
 # `dns-test` will invoke `docker build` from a temporary directory that contains
 # a clone of the hickory repository. `./src` here refers to that clone; not to
 # any directory inside the `hickory-dns` repository
+
+FROM chef AS planner
 COPY ./src /usr/src/hickory
-RUN --mount=type=cache,target=/usr/src/hickory/target \
-    cargo build --manifest-path /usr/src/hickory/Cargo.toml -p hickory-dns --features recursor,$DNSSEC_FEATURE && \
-    cargo build --manifest-path /usr/src/hickory/Cargo.toml --bin dns --features h3-aws-lc-rs,https-aws-lc-rs && \
-    cp /usr/src/hickory/target/debug/hickory-dns /usr/bin/ && \
-    cp /usr/src/hickory/target/debug/dns /usr/bin/
+WORKDIR /usr/src/hickory
+RUN cargo chef prepare
+
+FROM chef AS builder
+COPY --from=planner /usr/src/hickory/recipe.json /usr/src/hickory/recipe.json
+WORKDIR /usr/src/hickory
+RUN cargo chef cook -p hickory-dns --bin hickory-dns --features recursor,$DNSSEC_FEATURE && \
+    cargo chef cook -p hickory-util --bin dns --features h3-aws-lc-rs,https-aws-lc-rs
+COPY ./src /usr/src/hickory
+RUN cargo build -p hickory-dns --bin hickory-dns --features recursor,$DNSSEC_FEATURE && \
+    cargo build -p hickory-util --bin dns --features h3-aws-lc-rs,https-aws-lc-rs
+
+FROM debian:bookworm-slim AS final
+# ldns-utils = ldns-{key2ds,keygen,signzone}
+RUN apt-get update && \
+    apt-get install -y \
+    ldnsutils \
+    bind9-utils \
+    tshark \
+    openssl \
```

Comment:
Why does this additionally need openssl? Maybe a comment for each dependency on how/why it is used?

Reply:
The `rust:1-slim-bookworm` image installs `openssl`, but `debian:bookworm-slim` does not, and we need it to generate a keypair for Hickory to sign zones with when running as an authoritative name server.